pkg/controller/operators/catalogtempate/operator.go

```diff
@@ -0,0 +1,311 @@
+
+	// Wire CatalogSources
+	catsrcInformer := crInformerFactory.Operators().V1alpha1().CatalogSources()
+	op.lister.OperatorsV1alpha1().RegisterCatalogSourceLister(metav1.NamespaceAll, catsrcInformer.Lister())
+	catalogTemplateSrcQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "catalogSourceTemplate")
+	op.catalogSourceTemplateQueueSet.Set(metav1.NamespaceAll, catalogTemplateSrcQueue)
+	catsrcQueueInformer, err := queueinformer.NewQueueInformer(
+		op.ctx,
+		// TODO: commented out sections I don't think are necessary
+		// queueinformer.WithMetricsProvider(metrics.NewMetricsCatalogSource(op.lister.OperatorsV1alpha1().CatalogSourceLister())),
+		queueinformer.WithLogger(op.logger),
+		queueinformer.WithQueue(catalogTemplateSrcQueue),
+		queueinformer.WithInformer(catsrcInformer.Informer()),
+		queueinformer.WithSyncer(queueinformer.LegacySyncHandler(op.syncCatalogSources).ToSyncer()), // ToSyncerWithDelete(op.handleCatSrcDeletion)), TODO do we need to handle deletion specially?
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := op.RegisterQueueInformer(catsrcQueueInformer); err != nil {
+		return nil, err
+	}
+
+	return op, nil
+}
+
+func (o *Operator) syncCatalogSources(obj interface{}) error {
+	// this is an opportunity to update the server version (regardless of any other actions for processing a catalog source)
+	o.updateServerVersion()
+
+	inputCatalogSource, ok := obj.(*v1alpha1.CatalogSource)
+	if !ok {
+		o.logger.Debugf("wrong type: %#v", obj)
+		return fmt.Errorf("casting CatalogSource failed")
+	}
+
+	outputCatalogSource := inputCatalogSource.DeepCopy()
+
+	logger := o.logger.WithFields(logrus.Fields{
+		"catSrcName": outputCatalogSource.GetName(),
+		"id":         queueinformer.NewLoopID(),
+	})
+	logger.Info("syncing catalog source for annotation templates")
+
+	// this is our opportunity to discover GVK templates and setup watchers (if possible)
+	foundGVKs := catalogsource.InitializeCatalogSourceTemplates(outputCatalogSource)
+	for _, gvk := range foundGVKs {
+		o.processGVK(o.ctx, logger, gvk.GroupVersionKind)
+	}
+
+	catalogImageTemplate := catalogsource.GetCatalogTemplateAnnotation(outputCatalogSource)
+	if catalogImageTemplate == "" {
+		logger.Info("this catalog source is not participating in template replacement")
+		// make sure the conditions are removed
+		catalogsource.RemoveStatusConditions(logger, o.client, outputCatalogSource, StatusTypeTemplatesHaveResolved, StatusTypeResolvedImage)
+		// no further action is needed
+		return nil
+	}
+
+	processedCatalogImageTemplate, unresolvedTemplates := catalogsource.ReplaceTemplates(catalogImageTemplate)
+
+	templatesAreResolved := len(unresolvedTemplates) == 0
+	// curly braces are not allowed in a real image reference (see https://github.com/distribution/distribution/blob/2461543d988979529609e8cb6fca9ca190dc48da/reference/reference.go#L4-L24)
+	invalidCharsFound := strings.ContainsAny(processedCatalogImageTemplate, "{}")
```

Comment:
can we use an image distribution lib to validate the string? I assume there are other potentially invalid values here, or at the very least we want to separate the templating bits out and potentially validate afterwards?

Reply:
I am familiar with that code and have used it elsewhere. Should be an easy thing to add... frankly I should have thought of this.

Reply:
added the distribution lib and used its parser